scrypto-test/src/environment/builder.rs

```diff
@@ -0,0 +1,532 @@
+            database: InMemorySubstateDatabase::standard(),
+            flash_database: FlashSubstateDatabase::standard(),
+            added_global_references: Default::default(),
+            bootstrap: true,
+            protocol_updates: ProtocolUpdateOptIns::default(),
+        }
+    }
+}
+
+impl<D> TestEnvironmentBuilder<D>
+where
+    D: SubstateDatabase + CommittableSubstateDatabase + 'static,
+{
+    const DEFAULT_INTENT_HASH: Hash = Hash([0; 32]);
+
+    pub fn flash(mut self, data: DbFlash) -> Self {
+        // Flash the substates to the database.
+        let database_updates = DatabaseUpdates {
+            node_updates: data
+                .into_iter()
+                .map(|(db_node_key, partition_num_to_updates_mapping)| {
+                    (
+                        db_node_key,
+                        NodeDatabaseUpdates {
+                            partition_updates: partition_num_to_updates_mapping
+                                .into_iter()
+                                .map(|(partition_num, substates)| {
+                                    (
+                                        partition_num,
+                                        PartitionDatabaseUpdates::Delta {
+                                            substate_updates: substates
+                                                .into_iter()
+                                                .map(|(db_sort_key, value)| {
+                                                    (db_sort_key, DatabaseUpdate::Set(value))
+                                                })
+                                                .collect(),
+                                        },
+                                    )
+                                })
+                                .collect(),
+                        },
+                    )
+                })
+                .collect(),
+        };
+        self.flash_database.commit(&database_updates);
+
+        self
+            /* Global references found in the NodeKeys */
+            .add_global_references(
+                database_updates
+                    .node_updates
+                    .keys()
+                    .map(SpreadPrefixKeyMapper::from_db_node_key)
+                    .filter_map(|item| GlobalAddress::try_from(item).ok()),
+            )
+            /* Global references found in the Substate Values */
+            .add_global_references(
+                database_updates
+                    .node_updates
+                    .values()
+                    .flat_map(|NodeDatabaseUpdates { partition_updates }| {
+                        partition_updates.values()
+                    })
+                    .filter_map(|item| {
+                        if let PartitionDatabaseUpdates::Delta { substate_updates } = item {
+                            Some(substate_updates.values())
+                        } else {
+                            None
```

Comment:
Should handle `PartitionDatabaseUpdates::Reset` too? Alternatively, we don't use `PartitionDatabaseUpdates` enum since no partition deletion seems to be supported.

Reply:
Good point on handling reset too, I missed that. I've handled it and updated the PR.